upload/src/addons/SV/WarningImprovements/XF/Entity/WarningDefinition.php

```diff
@@ -16,6 +16,35 @@
  */
 class WarningDefinition extends XFCP_WarningDefinition
 {
+    public function canView(&$error = null)
+    {
+        $visitor = \XF::visitor();
+
+        if (empty($visitor->user_id))
+        {
+            return false;
+        }
+
+        return $this->isUsableByUser($visitor, $error);
+    }
+
+    public function isUsableByUser(\XF\Entity\User $user = null, &$error = null)
+    {
+        $user = $user ?: \XF::visitor();
+
+        foreach ($this->Category->allowed_user_group_ids AS $userGroupId)
+        {
+            if ($userGroupId == -1 || $user->isMemberOf($userGroupId))
+            {
+                return true;
```

Comment:
The alternative is a top-down prune based off using `$categoryTree` in the caller. This is probably the best, as all the categories are loaded at this stage, and  it removes large chunks at a time rather than per-warning